src/source-control/GitSourceControl.ts

```diff
@@ -265,64 +265,70 @@ export class GitSourceControl implements SourceControl {
+          repo.detachHead(); // Be sure to set the head after detaching!
+        }
+
+        const newBranchRef = await repo.createBranch(
+          localRefToBranchName(refName),
+          newCommitOid,
+          true,
+        );
+
+        if (isBranchHead) {
+          // If we were checked out to the pre-cherry-pick branch, check out the
+          // new post-cherry-pick branch.
+          repo.setHead(newBranchRef.toString());
         }
       });
-      queue.push(...baseCommitSttack.childCommits);
 
       // Update refs in our hashMap
-      this.repo.commits = produce(this.repo.commits, (draftCommitHashMap) => {
-        const childrenOfTargetCommit = targetCommitSttack.childCommits;
+      this.repo = produce(this.repo, (draftRepo) => {
+        const draftTargetCommit = draftRepo.commits.get(targetCommitHash)!;
+        const childrenOfTargetCommit = draftTargetCommit.childCommits;
+
+        // Add the newly-created commit to our repo
         const newCommitToInsert: Commit = {
           hash: newCommitOid.tostrS(),
-          title: "Sttack Commit",
+          title: commitToBeRebased.title,
           timestamp: new Date(),
-          author: baseCommitSttack.author,
-          committer: baseCommitSttack.committer,
-          refNames: baseCommitSttack.refNames.filter(refIsLocal),
-          parentCommits: [targetCommitSttack.hash],
-          childCommits: [...childrenOfTargetCommit],
+          author: commitToBeRebased.author,
+          committer: commitToBeRebased.committer,
+          refNames: commitToBeRebased.refNames.filter(refIsLocal),
+          parentCommits: [draftTargetCommit.hash],
+          childCommits: [],
         };
-        draftCommitHashMap.set(newCommitOid.tostrS(), newCommitToInsert);
-        // Update ChildCommitHash for targetCommitSttack to newCommit
-        targetCommitSttack.childCommits = [newCommitOid.tostrS()];
-        // Update ParentCommitHash for all of the children and replace targetCommitSttack's hash with newCommit's hash
-        childrenOfTargetCommit.forEach((child) => {
-          const childCommit = draftCommitHashMap.get(child)!;
-          const parentsOfChild = childCommit.parentCommits.map((parent) => {
-            if (parent === targetCommitSttack.hash) {
-              return newCommitOid.tostrS();
-            }
-            return parent;
-          });
-          childCommit.parentCommits = parentsOfChild;
-          draftCommitHashMap.set(child, childCommit);
-        });
+        draftRepo.commits.set(newCommitOid.tostrS(), newCommitToInsert);
+
+        // Update target's children
+        draftTargetCommit.childCommits = [
+          ...draftTargetCommit.childCommits,
+          newCommitOid.tostrS(),
+        ];
       });
+
+      queue.push(...commitToBeRebased.childCommits);
+
+      // We now would want to use this new commit hash as the target
+      // FIXME: This will turn a commit tree into a linear stack!
```

Comment:
So turns out this causes a real bug:

1. Have some commits on master: 
![image](https://user-images.githubusercontent.com/12784593/90314883-b32e8b00-df49-11ea-8298-b503f84a4117.png)
1. Rebase C (`c265a2`) onto B (`0952ea`). This works: a new C is created and all the local branches are moved onto it.
![image](https://user-images.githubusercontent.com/12784593/90314906-dbb68500-df49-11ea-8b43-97e114934398.png)
1. Rebase B (`0952ea`) onto A (`54c20d`)
	
	The UI right before pressing `c` to confirm the rebase:
	
	![image](https://user-images.githubusercontent.com/12784593/90314936-04d71580-df4a-11ea-800d-3b817ffe0a91.png)

	After rebase:

	![image](https://user-images.githubusercontent.com/12784593/90314966-318b2d00-df4a-11ea-87bf-27a866291562.png)

	2 new "C" commits are created. This is likely because line 325 below linearized the tree, applying both the local and remote C commits (`3fb693` and `c265a2`) on each other instead of applying both onto B.

This bug should also be reproducible when you rebase any tree, not just trees with remote branches.

Since this may be a pretty big issue, I think it's fine if we just chuck this into a new issue and merge this PR first.

Reply:
Opened #59 to track this